### PR TITLE
[AMDGPU] Don't send DEALLOC_VGPRs after calls

### PR DESCRIPTION
**Pull request by rovka**
Calls do not have to wait for VsCnt, so after they return there might still be scratch stores in progress. It's important that we don't send the DEALLOC_VGPR message in that case, since that might release the VGPRs and scratch allocation before those stores are complete.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 1]
```diff
@@ -1487,6 +1488,9 @@ void SIInsertWaitcnts::updateEventWaitcntAfter(MachineInstr &Inst,
     if (callWaitsOnFunctionReturn(Inst)) {
       // Act as a wait on everything
       ScoreBrackets->applyWaitcnt(AMDGPU::Waitcnt::allZeroExceptVsCnt());
+      // Since there's no guaranteed wait on VsCnt, scratch stores might still
+      // be in flight.
+      ScoreBrackets->setPendingEvent(SCRATCH_WRITE_ACCESS);
```
Comment (jasilvanus) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:1493]: Nit: There could also be in-flight global non-scratch stores, correct?

The comment on `SCRATCH_WRITE_ACCESS` says `vector-memory write that may be scratch` so these should be covered, but I wonder whether we should still explicitly include `VMEM_WRITE_ACCESS`.

Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:1493]: I suggest calling `setNonKernelFunctionInitialState` here (and renaming it to something like `setStateOnFunctionEntryOrReturn`).

**Review by jayfoad — APPROVED** [review]
LGTM

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp (hunk 1 above)
Comment (rovka) [reply llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp]: > Nit: There could also be in-flight global non-scratch stores, correct?
> 
> The comment on `SCRATCH_WRITE_ACCESS` says `vector-memory write that may be scratch` so these should be covered, but I wonder whether we should still explicitly include `VMEM_WRITE_ACCESS`.

I changed it to incorporate Jay's suggestion, so I think we're covered now (setStateOnFunctionEntryOrReturn doesn't care about the event type, it only deals in terms of VsCnt).

File: llvm/test/CodeGen/AMDGPU/release-vgprs.mir [hunk 2]
```diff
@@ -565,3 +566,19 @@ body:             |
...
+  bb.0:
+    ; Make sure we don't send DEALLOC_VGPRS after a call, since there might be
+    ; scratch stores still in progress.
+    ; CHECK-LABEL: name: with_calls
+    ; CHECK-NOT: S_SENDMSG 3
+    ; CHECK: S_ENDPGM 0
+    GLOBAL_STORE_DWORD undef renamable $vgpr0_vgpr1, killed renamable $vgpr1, 0, 4, implicit $exec
+    $sgpr30_sgpr31 = SI_CALL undef renamable $sgpr4_sgpr5, 0, csr_amdgpu
+    S_ENDPGM 0
+...
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/release-vgprs.mir:584]: is it worth adding a tail call case too?

File: llvm/test/CodeGen/AMDGPU/release-vgprs.mir [hunk 3]
```diff
@@ -582,3 +583,19 @@ body:             |
...
+body:             |
+  bb.0:
+    ; Make sure we don't send DEALLOC_VGPRS after a tail call, since there might be
+    ; scratch stores still in progress.
+    ; CHECK-LABEL: name: with_tail_calls
+    ; CHECK-NOT: S_SENDMSG 3
+    ; CHECK: S_ENDPGM 0
+    GLOBAL_STORE_DWORD undef renamable $vgpr0_vgpr1, killed renamable $vgpr1, 0, 4, implicit $exec
+    SI_TCRETURN undef renamable $sgpr4_sgpr5, @with_tail_calls, 0, csr_amdgpu
+    S_ENDPGM 0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/release-vgprs.mir:600]: This wouldn't have an S_ENDPGM after it

File: llvm/test/CodeGen/AMDGPU/release-vgprs.mir [hunk 4]
```diff
@@ -23,6 +23,7 @@
   define amdgpu_ps void @image_atomic() { ret void }
   define amdgpu_ps void @global_store_optnone() noinline optnone { ret void }
   define amdgpu_cs void @with_calls() { ret void }
+  define amdgpu_cs void @with_tail_calls() { ret void }
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/release-vgprs.mir:26]: entry functions can't end in tail calls?

File: llvm/test/CodeGen/AMDGPU/release-vgprs.mir (hunk 3 above)
Comment (rovka) [reply llvm/test/CodeGen/AMDGPU/release-vgprs.mir:600]: Oops, I didn't think that through. If there's no S_ENDPGM, we would never try to send DEALLOC_VGPRS in the first place. But I can update the test anyway, just so we can have this case covered for the future.

**Review by arsenm — APPROVED** [review]

**Pull request merged**